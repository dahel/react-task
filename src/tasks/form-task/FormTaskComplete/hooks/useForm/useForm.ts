import { useState, SyntheticEvent, ReactNode } from 'react';
import type { ISelectBaseOptions } from '../../components/selectBase/SelectBase';

export interface IValidator {
  validate: (value: string) => boolean;
  message: string;
}

export interface IFormItemConfig {
  name: string;
  value: string;
  label: string;
  render: RenderFormItemMethod;
  placeholder?: string;
  required?: boolean;
  requiredErrorMessage?: string;
  validationError?: string;
  valid?: boolean;
  disabled?: boolean;
  options?: ISelectBaseOptions[];
  validation?: IValidator[];
  touched?: boolean;
  update?: UpdateFormItemMethod;
}

export interface IFormConfig {
  [key: string]: IFormItemConfig;
}

export interface IFormValues {
  [key: string]: string;
}

type RenderFormItemMethod = (formItemConfig: IFormItemConfig, callback: (event: SyntheticEvent) => void) => ReactNode;

type UpdateFormItemMethod = (formConfig: IFormConfig, value: string) => IFormConfig;

interface IValidationResult {
  valid: boolean;
  validationError: string;
}

export type Form = { [key: string]: { render: () => ReactNode } };

const getValidationResult = (valid: boolean, validationError: string = ''): IValidationResult => ({
  valid,
  validationError,
});

const validate = (value: string, formConfigField: IFormItemConfig): IValidationResult => {
  if (formConfigField.disabled) {
    return getValidationResult(true);
  }

  if (!formConfigField.required && !value.length) {
    return getValidationResult(true);
  }

  if (formConfigField.required && !value.length) {
    return getValidationResult(false, formConfigField.requiredErrorMessage || 'Field required');
  }

  const failedValidator = (formConfigField.validation || []).find((validator) => !validator.validate(value));

  if (failedValidator) {
    return getValidationResult(false, failedValidator.message);
  }

  return getValidationResult(true);
};

export default function useForm(
  formFields: IFormConfig,
): [Form, () => IFormValues, () => boolean, (formConfig: any) => void] {
  const [formConfig, setFormConfig] = useState(formFields);
  const handleInputEvent = (event: SyntheticEvent): void => {
    const { name, value } = event.target as HTMLInputElement | HTMLSelectElement;
    const update: UpdateFormItemMethod = formConfig[name].update || ((formConfig: IFormConfig) => formConfig);
    const { valid, validationError }: IValidationResult = validate(value, formConfig[name]);

    setFormConfig(
      update(
        {
          ...formConfig,
          [name]: {
            ...formConfig[name],
            value,
            valid,
            validationError,
            touched: true,
          },
        },
        value,
      ),
    );
  };

  const form: Form = Object.keys(formConfig).reduce((result: Form, formItemKey: string) => {
    const formItemConfig: IFormItemConfig = formConfig[formItemKey];

    return {
      ...result,
      [formItemKey]: {
        render() {
          return formItemConfig.render(formItemConfig, handleInputEvent);
        },
      },
    };
  }, {} as Form);

  const getValues = (): IFormValues => {
    return Object.values(formConfig).reduce(
      (result: IFormValues, next: IFormItemConfig) =>
        ({
          ...result,
          ...(next.value.trim() ? { [next.name]: next.value.trim() } : null),
        } as IFormValues),
      {} as IFormValues,
    );
  };

  const validateForm = (): boolean => {
    const { updatedFormConfig, isFormValid }: { updatedFormConfig: IFormConfig; isFormValid: boolean } = Object.keys(
      formConfig,
    ).reduce(
      (result: { updatedFormConfig: IFormConfig; isFormValid: boolean }, nextKey: string) => {
        const { valid, validationError } = validate(formConfig[nextKey].value, formConfig[nextKey]);

        return {
          updatedFormConfig: {
            ...result.updatedFormConfig,
            [nextKey]: {
              ...formConfig[nextKey],
              valid,
              validationError,
              touched: true,
            },
          },
          isFormValid: result.isFormValid && valid,
        };
      },
      { updatedFormConfig: {}, isFormValid: true } as {
        updatedFormConfig: IFormConfig;
        isFormValid: boolean;
      },
    );

    setFormConfig(updatedFormConfig);

    return isFormValid;
  };

  const reset = (formConfig: any) => setFormConfig(formConfig);

  return [form, getValues, validateForm, reset];
}
