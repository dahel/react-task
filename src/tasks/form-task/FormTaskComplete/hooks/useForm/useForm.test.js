import renderer from 'react-test-renderer';
import { renderHook, act } from '@testing-library/react-hooks';
import useForm from './useForm';

let formFields;

beforeEach(() => {
  formFields = {
    test_field_1: {
      name: 'test_field_1-name',
      label: '',
      value: '',
      required: true,
      requiredErrorMessage: 'requiredErrorMessage',
      render: jest.fn(() => <div>test_field_1 element</div>),
      validation: [
        {
          validate: () => {},
          message: 'Invalid length',
        },
      ],
    },
    test_field_2: {
      name: 'test_field_2-name',
      label: '',
      value: '',
      render: jest.fn(() => <div>test_field_2 element</div>),
      validation: [
        {
          validate: () => {},
          message: 'Invalid format',
        },
      ],
    },
    test_field_3: {
      name: 'test_field_3-name',
      label: '',
      value: '',
      render: jest.fn(() => <div>test_field_3 element</div>),
      update: () => {},
    },
  };
});

test('generates form entry for every form field', () => {
  const { result } = renderHook(() => useForm(formFields));
  const [form] = result.current;

  expect(form.test_field_1).toBeDefined();
  expect(typeof form.test_field_1.render).toBe('function');

  expect(form.test_field_2).toBeDefined();
  expect(typeof form.test_field_2.render).toBe('function');

  expect(form.test_field_3).toBeDefined();
  expect(typeof form.test_field_3.render).toBe('function');
});

test('properly renders every field item', () => {
  const { result } = renderHook(() => useForm(formFields));
  const [form] = result.current;

  expect(renderer.create(form.test_field_1.render()).toJSON()).toMatchSnapshot();
  expect(renderer.create(form.test_field_2.render()).toJSON()).toMatchSnapshot();
  expect(renderer.create(form.test_field_3.render()).toJSON()).toMatchSnapshot();
});

test(`calls 'update' function of form item config after item's value has change`, () => {
  const updateSpy = jest.fn();
  let cachedOnInputChange;

  formFields.test_field_1.required = true;
  formFields.test_field_1.value = 'asdf';
  formFields.test_field_1.update = updateSpy;
  formFields.test_field_1.render = (config, onInputChange) => {
    cachedOnInputChange = onInputChange;

    return <div>test_field_1 element</div>;
  };

  const { result } = renderHook(() => useForm(formFields));
  const [form] = result.current;

  form.test_field_1.render();

  act(() => {
    cachedOnInputChange({
      target: { value: 'new value', name: 'test_field_1' },
    });
  });

  expect(updateSpy).toHaveBeenCalledWith(expect.any(Object), 'new value');
});

describe('getValues', () => {
  test(`returns properly formatted object`, () => {
    formFields.test_field_1.value = 'one';
    formFields.test_field_2.value = 'two';
    formFields.test_field_3.value = 'three';

    const { result } = renderHook(() => useForm(formFields));
    const [, getValues] = result.current;

    expect(getValues()).toEqual({
      'test_field_1-name': 'one',
      'test_field_2-name': 'two',
      'test_field_3-name': 'three',
    });
  });
});

describe('validateForm', () => {
  test(`calls validate function of every validation config after validateForm is called and none of required fields is empty`, () => {
    const validateSpy1 = jest.fn();
    const validateSpy2 = jest.fn();

    formFields.test_field_1.value = 'asdf';
    formFields.test_field_1.validation[0] = { validate: validateSpy1 };
    formFields.test_field_2.value = 'asdf';
    formFields.test_field_2.validation[0] = { validate: validateSpy2 };

    const { result } = renderHook(() => useForm(formFields));
    const [, , validateForm] = result.current;

    act(() => {
      validateForm();
    });

    expect(validateSpy1).toHaveBeenCalled();
    expect(validateSpy2).toHaveBeenCalled();
  });

  test(`returns true in case all fields are required and have defined values and have no validators`, () => {
    let isFormValid;

    formFields.test_field_1.required = true;
    formFields.test_field_1.value = 'asdf';
    formFields.test_field_1.validation = [];
    formFields.test_field_2.required = true;
    formFields.test_field_2.value = 'asdf';
    formFields.test_field_2.validation = [];

    const { result } = renderHook(() => useForm(formFields));
    const [, , validateForm] = result.current;

    act(() => {
      isFormValid = validateForm();
    });

    expect(isFormValid).toBe(true);
  });

  test(`returns true in case all fields are required and have defined values and their validators returns true`, () => {
    let isFormValid;

    formFields.test_field_1.required = true;
    formFields.test_field_1.value = 'asdf';
    formFields.test_field_1.validation = [{ validate: () => true }];
    formFields.test_field_2.required = true;
    formFields.test_field_2.value = 'asdf';
    formFields.test_field_2.validation = [{ validate: () => true }];

    const { result } = renderHook(() => useForm(formFields));
    const [, , validateForm] = result.current;

    act(() => {
      isFormValid = validateForm();
    });

    expect(isFormValid).toBe(true);
  });

  test(`returns false in case all fields are not required and on of the validator returns false`, () => {
    let isFormValid;

    formFields.test_field_1.required = true;
    formFields.test_field_1.value = 'asdf';
    formFields.test_field_1.validation = [{ validate: () => true }];
    formFields.test_field_2.required = true;
    formFields.test_field_2.value = 'asdf';
    formFields.test_field_2.validation = [{ validate: () => false }];

    const { result } = renderHook(() => useForm(formFields));
    const [, , validateForm] = result.current;

    act(() => {
      isFormValid = validateForm();
    });

    expect(isFormValid).toBe(false);
  });
});
