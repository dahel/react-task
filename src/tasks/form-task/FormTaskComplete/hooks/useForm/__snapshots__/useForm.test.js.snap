// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useForm hook properly renders every field item 1`] = `
<div>
  test_field_1 element
</div>
`;

exports[`useForm hook properly renders every field item 2`] = `
<div>
  test_field_2 element
</div>
`;

exports[`useForm hook properly renders every field item 3`] = `
<div>
  test_field_3 element
</div>
`;
